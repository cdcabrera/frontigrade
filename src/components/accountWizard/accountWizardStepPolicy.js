import React from 'react';
import PropTypes from 'prop-types';
import { FieldLevelHelp, Form, Grid } from 'patternfly-react';
import helpers from '../../common/helpers';
import { connect, reduxTypes, store } from '../../redux';
import apiTypes from '../../constants/apiConstants';
import { FormField, fieldValidation } from '../formField/formField';
import CopyField from '../copyField/copyField';

// FixMe: API - Hard Coded Values - "AwsAccount"
class AccountWizardStepPolicy extends React.Component {
  state = {
    accountName: '',
    accountNameError: null,
    resourceType: 'AwsAccount'
  };

  onChangeAccountName = event => {
    const { value } = event.target;
    const errorMessage = fieldValidation.doesntHaveMinimumCharacters(value, 1) ? 'You must name your account' : '';

    this.setState(
      {
        accountName: value,
        accountNameError: errorMessage
      },
      () => this.isStepValid()
    );
  };

  // ToDo: when form valid evaluate moving to next step onEnter key-press event
  onSubmit = event => {
    event.preventDefault();
  };

  isStepValid() {
    const { accountName, accountNameError, resourceType } = this.state;
    const stepValid = accountNameError === '';
    const dispatchType = stepValid
      ? reduxTypes.account.ADD_ACCOUNT_WIZARD_STEP_POLICY
      : reduxTypes.account.INVALID_ACCOUNT_WIZARD_STEP_POLICY;

    store.dispatch({
      type: dispatchType,
      account: {
        [apiTypes.API_SUBMIT_ACCOUNT_NAME]: accountName,
        [apiTypes.API_SUBMIT_ACCOUNT_RESOURCE_TYPE]: resourceType
      }
    });
  }

  render() {
    const { accountName, accountNameError, resourceType } = this.state;
    const { awsConfigPolicySetup, stepPolicyValid, stepPolicyErrorMessage } = this.props;

    let stepError = null;

    if (!stepPolicyValid) {
      stepError = stepPolicyErrorMessage;
    }

    let displayAwsConfigPolicySetup = awsConfigPolicySetup;

    if (displayAwsConfigPolicySetup) {
      displayAwsConfigPolicySetup = helpers.prettyPrintJson(displayAwsConfigPolicySetup);
    }

    const popover = <p>The name or alternative name for the account. Account names must be unique.</p>;

    const nameLabel = (
      <div className="cloudmeter-label-nowrap">
        Account Name <FieldLevelHelp content={popover} close />
      </div>
    );

    return (
      <Form horizontal onSubmit={this.onSubmit}>
        <FormField label={nameLabel} error={stepError || accountNameError} errorMessage={stepError || accountNameError}>
          <Form.FormControl
            autoFocus
            type="text"
            name="accountName"
            value={accountName}
            maxLength={256}
            onChange={this.onChangeAccountName}
          />
        </FormField>
        <Form.FormControl type="hidden" name="resourceType" value={resourceType} />
        <Form.FormGroup>
          <Grid.Col sm={12} className="padding-left">
            <ul>
              <li>
                Sign in to the{' '}
                <a href="https://console.aws.amazon.com/iam" target="_blank" rel="noopener noreferrer">
                  AWS Identity Access Management
                </a>{' '}
                (IAM) console.
              </li>
              <li>
                <p>Create a new policy, pasting the following content into the JSON text box.</p>
                <CopyField multiline value={displayAwsConfigPolicySetup} />
              </li>
              <li>Complete the process to create your new policy.</li>
            </ul>
          </Grid.Col>
        </Form.FormGroup>
      </Form>
    );
  }
}

AccountWizardStepPolicy.propTypes = {
  awsConfigPolicySetup: PropTypes.object,
  stepPolicyValid: PropTypes.bool,
  stepPolicyErrorMessage: PropTypes.string
};

AccountWizardStepPolicy.defaultProps = {
  awsConfigPolicySetup: {},
  stepPolicyValid: false,
  stepPolicyErrorMessage: null
};

const mapStateToProps = state => ({ ...state.accountWizard });

const ConnectedAccountWizardStepPolicy = connect(mapStateToProps)(AccountWizardStepPolicy);

export { ConnectedAccountWizardStepPolicy as default, ConnectedAccountWizardStepPolicy, AccountWizardStepPolicy };
