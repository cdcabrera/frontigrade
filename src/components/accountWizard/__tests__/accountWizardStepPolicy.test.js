import React from 'react';
import configureMockStore from 'redux-mock-store';
import { shallow } from 'enzyme';
import { ConnectedAccountWizardStepPolicy } from '../accountWizardStepPolicy';

describe('AccountWizardStepPolicy Component', () => {
  const generateEmptyStore = (obj = {}) => configureMockStore()(obj);

  it('should render a wizard policy step', () => {
    const store = generateEmptyStore({ accountWizard: {} });
    const component = shallow(<ConnectedAccountWizardStepPolicy />, { context: { store } });

    expect(component.dive()).toMatchSnapshot('connected');
  });
});
