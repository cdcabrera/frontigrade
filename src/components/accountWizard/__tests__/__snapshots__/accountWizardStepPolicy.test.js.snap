// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountWizardStepPolicy Component should render a wizard policy step: connected 1`] = `
<Form
  bsClass="form"
  componentClass="form"
  horizontal={true}
  inline={false}
  onSubmit={[Function]}
>
  <FormField
    colField={9}
    colFieldClassName={null}
    colLabel={3}
    colLabelClassName={null}
    error={null}
    errorMessage={null}
    id={null}
    label={
      <div
        className="cloudmeter-label-nowrap"
      >
        Account Name 
        <FieldLevelHelp
          buttonClass=""
          close={true}
          content={
            <p>
              The name or alternative name for the account. Account names must be unique.
            </p>
          }
          placement="top"
          rootClose={true}
        />
      </div>
    }
  >
    <FormControl
      autoFocus={true}
      bsClass="form-control"
      componentClass="input"
      maxLength={256}
      name="accountName"
      onChange={[Function]}
      type="text"
      value=""
    />
  </FormField>
  <FormControl
    bsClass="form-control"
    componentClass="input"
    name="resourceType"
    type="hidden"
    value="AwsAccount"
  />
  <FormGroup
    bsClass="form-group"
  >
    <Col
      bsClass="col"
      className="padding-left"
      componentClass="div"
      sm={12}
    >
      <ul>
        <li>
          Sign in to the
           
          <a
            href="https://console.aws.amazon.com/iam"
            rel="noopener noreferrer"
            target="_blank"
          >
            AWS Identity Access Management
          </a>
           
          (IAM) console.
        </li>
        <li>
          <p>
            Create a new policy, pasting the following content into the JSON text box.
          </p>
          <CopyField
            expandDescription={null}
            id={null}
            label={
              <Icon
                name="paste"
                type="fa"
              />
            }
            labelClicked={
              <Icon
                name="check"
                type="fa"
              />
            }
            labelClickedDescription="Copied"
            labelDescription="Copy to Clipboard"
            multiline={true}
            resetTimer={8000}
            tooltipPlacement="top"
            value="{}"
          />
        </li>
        <li>
          Complete the process to create your new policy.
        </li>
      </ul>
    </Col>
  </FormGroup>
</Form>
`;
